decimal-macros/src/lib.rs

```diff
@@ -0,0 +1,85 @@
+#![feature(plugin_registrar, rustc_private)]
+
+extern crate libc;
+#[doc(hidden)]
+pub extern crate decimal;
+
+extern crate rustc_plugin;
+extern crate syntax;
+
+use rustc_plugin::Registry;
+use syntax::ext::base::{DummyResult, MacEager, ExtCtxt, MacResult};
+use syntax::ext::build::AstBuilder;
+use syntax::ext::source_util;
+use syntax::codemap::Span;
+use syntax::ast::{ExprKind, TokenTree, LitKind, StrStyle};
+
+use decimal::d128;
+
+fn d128_lit<'cx>(cx: &'cx mut ExtCtxt, sp: Span, tts: &[TokenTree]) -> Box<MacResult + 'cx> {
+    let mac_res = source_util::expand_stringify(cx, sp, tts);
+    
+    let ex = match mac_res.make_expr() {
+        Some(ex) => ex,
+        None => {
+            cx.span_err(sp, "one argument needed");
+            return DummyResult::any(sp);
+        }
+    };
+    
+    match &ex.node {
+        &ExprKind::Lit(ref lit) => match &lit.node {
+            &LitKind::Str(ref s, StrStyle::Cooked) => {
+                if (*s).len() == 0 {
+                    cx.span_err(sp, "one argument needed");
+                    return DummyResult::any(sp);
+                }
+                let num = match from_str(&*s) {
+                    Ok(num) => num,
+                    Err(s) => {
+                        cx.span_err(lit.span, s);
+                        return DummyResult::any(sp);
+                    }
+                };
+                let num = unsafe { ::std::mem::transmute::<d128, [u8; 16]>(num) };
+                
+                let mut vec = Vec::new();
+                for i in 0..16 {
+                    vec.push(cx.expr_u8(lit.span, num[i]));
+                }
+                let vec = cx.expr_vec(lit.span, vec);
+                let ids = vec![cx.ident_of("decimal"), cx.ident_of("d128"), cx.ident_of("from_bytes")];
+                let ex = cx.expr_call_global(lit.span, ids, vec![vec]);
+
+                return MacEager::expr(ex);
+            },
+            _ => {}
```

Comment:
The `source_util::expand_stringify` at the beginning ensures that the item is a string literal. I chose this because it's the easiest way to parse the number: Just defer it to the function in `d128`.